ios/NedaaWidget/PrayerCountdownView.swift

```diff
@@ -0,0 +1,168 @@
+    typealias Entry = PrayerEntry
+    typealias Intent = ConfigurationIntent
+    
+    func placeholder(in context: Context) -> PrayerEntry {
+        PrayerEntry(date: Date(), configuration: ConfigurationIntent(), nextPrayer: PrayerData(name:NSLocalizedString("fajr", comment: "fajr"), date: Date().addingTimeInterval(TimeInterval(60 * 5))), previousPrayer: PrayerData(name: NSLocalizedString("isha", comment: "isha"), date: Date().addingTimeInterval(3600)) )
+    }
+    
+    func getSnapshot(for configuration: ConfigurationIntent, in context: Context, completion: @escaping (PrayerEntry) -> ()) {
+        let entry = PrayerEntry(date: Date(), configuration: configuration, nextPrayer: PrayerData(name: NSLocalizedString("fajr", comment: "fajr"), date: Date().addingTimeInterval(3600)), previousPrayer: PrayerData(name: NSLocalizedString("isha", comment: "Isha"), date: Date().addingTimeInterval(3600)))
+        completion(entry)
+    }
+    
+    func getTimeline(for configuration: ConfigurationIntent, in context: Context, completion: @escaping (Timeline<PrayerEntry>) -> ()) {
+        var _: [PrayerEntry] = []
+        let prayerService = PrayerDataService()
+        let showSunrise = configuration.showSunrise as! Bool?
+        let nextPrayer = prayerService.getNextPrayer(showSunrise: showSunrise ?? true) ?? PrayerData(name: "GET NOT WOKRING", date: Date())
+        let previousPrayer = prayerService.getPreviousPrayer(showSunrise: showSunrise ?? true) ?? PrayerData(name: "GET NOT WOKRING", date: Date())
+        let currentDate = Date()
+        
+        let nextUpdateDate = calculateNextUpdateDate(currentDate: currentDate, nextPrayerDate: nextPrayer.date, previousPrayerDate: previousPrayer.date)
+        
+        let entry = PrayerEntry(date: currentDate,configuration: configuration, nextPrayer: nextPrayer, previousPrayer: previousPrayer)
+        let timeline = Timeline(entries: [entry], policy: .after(nextUpdateDate))
+        let dateformat = DateFormatter()
+        dateformat.dateFormat = "h:mm a"
+        
+        completion(timeline)
+    }
+    
+    func calculateNextUpdateDate(currentDate: Date, nextPrayerDate: Date, previousPrayerDate: Date) -> Date {
+        let timeIntervalToNextPrayer = nextPrayerDate.timeIntervalSince(currentDate)
+        let timeIntervalSincePreviousPrayer = currentDate.timeIntervalSince(previousPrayerDate)
+        
+        if timeIntervalSincePreviousPrayer < 1800 {
+            // If the previous prayer was less than 30 minutes ago, update 30 minutes after the previous prayer
+            return previousPrayerDate.addingTimeInterval(1800)
+        } else if timeIntervalToNextPrayer > 3600 {
+            // If the next prayer is more than 1 hour away, update 60 minutes before the next prayer
+            return nextPrayerDate.addingTimeInterval(-3600)
+        } else {
+            // Otherwise, update 30 minutes after the next prayer
+            return nextPrayerDate.addingTimeInterval(1800)
+        }
```

Comment:
don't we want to update on the prayer itself? not sure if it will get an update on the prayer time itself, since we are updating before the prayer by 1h and then after the prayer by 30m

Reply:
Okay so we want to count up 30min after the prayer. We can schedule an update on the prayer itself then another 30min after for the count up. But, once the count down gets to zero it'll automatically start counting up as the date in the past this lets us achieve the required logic with one less update